File: python/hashing/hashMap.py
class HashMap:
    def __init__(self):
        self.chains = [None] * 1000


    class Chain:
        def __init__(self):
            self.head = None
            self.tail = None

        class Pair:
            def __init__(self, key, value):
                self.key = key
                self.value = value
                self.prev = None
                self.next = None

        def push(self, key, value):
            pair = self.Pair(key, value)
            if self.head is None:
                self.head = pair
            if self.tail is not None:
                self.tail.next = pair
                pair.prev = self.tail
            self.tail = pair

        def find(self, key):
            pair = self.head
            while pair is not None:
                if pair.key == key:
                    return pair
                pair = pair.next
            return None

        def pop(self, key):
            pair = self.find(key)

            if pair is None:
                return pair

            prev = pair.prev
            next = pair.next
            if prev is not None:
                prev.next = next
            else:
                self.head = next

            if next is not None:
                next.prev = prev
            else:
                self.tail = prev
            
            result = pair.value
            pair = None
            return result


    def hash(self, key):
        return int(key[:3])


    def hashKey(self, key):
        chain = self.chains[self.hash(key)]
        if chain is None:
            return False
        return (False if chain.find(key) is None else True)


    def set(self, key, value):
        chain = self.chains[self.hash(key)]

        if chain is None:
            self.chains[self.hash(key)] = chain = self.Chain()
        
        exist = chain.find(key)
        if exist is not None:
            exist.value = value
        else:
            chain.push(key, value)


    def get(self, key):
        chain = self.chains[self.hash(key)]

        if chain is None:
            return None

        result = chain.find(key)

        if result is None:
            return None
        else:
            return result.value

    def remove(self, key):
        chain = self.chains[self.hash(key)]

        if chain is None:
            return None

        return chain.pop(key)

File: java/src/hashing/HashMap.java
package hashing;

public class HashMap {

    private Chain[] chains = new Chain[1000];

    private class Chain {
        private Pair head;
        private Pair tail;

        public Pair find(String key) {
            Pair pair = this.head;
            while (pair != null) {
                if (pair.key == key)
                    return pair;
                pair = pair.next;
            }
            return null;
        }

        public void push(String key, String value) {
            Pair pair = new Pair(key, value);

            if (this.head == null)
                this.head = pair;

            if (this.tail != null) {
                this.tail.next = pair;
                pair.prev = this.tail;
            }

            this.tail = pair;
        }

        public String pop(String key) {
            Pair pair = this.find(key);
            if (pair == null)
                return null;

            Pair prev = pair.prev;
            Pair next = pair.next;

            if (prev != null)
                prev.next = next;
            else
                this.head = next;

            if (next != null)
                next.prev = prev;
            else
                this.tail = prev;

            String result = pair.value;
            pair = null;

            return result;
        }
    }

    private class Pair {
        private String key;

        private String value;

        private Pair next;

        private Pair prev;

        public Pair(String key, String value) {
            this.key = key;
            this.value = value;
        }
    }

    public int hash(String key) {
        return Integer.parseInt(key.substring(0, 3));
    }

    public void set(String key, String value) {
        Chain chain = this.chains[this.hash(key)];
        if (chain == null) {
            chain = new Chain();
            this.chains[this.hash(key)] = chain;
        }

        Pair result = chain.find(key);
        if (result == null)
            chain.push(key, value);
        else
            result.value = value;
    }

    public boolean hashKey(String key) {
        Chain chain = this.chains[this.hash(key)];

        return chain != null && chain.find(key) != null;
    }

    public String get(String key) {
        Chain chain = this.chains[this.hash(key)];

        if (chain == null)
            return null;

        Pair pair = chain.find(key);
        return pair != null ? pair.value : null;
    }

    public String remove(String key) {
        Chain chain = this.chains[this.hash(key)];

        if (chain == null)
            return null;

        return chain.pop(key);
    }
}
